code/parse/sexp.cpp

```diff
@@ -4298,173 +4356,189 @@ int get_sexp()
 
 		// Sexp string
 		else if (*Mp == '\"') {
+			auto startp = Mp;
 			auto len = strcspn(Mp + 1, "\"");
 			// was closing quote not found?
 			if (*(Mp + 1 + len) != '\"') {
-				Error(LOCATION, "Unexpected end of quoted string embedded in sexp!");
-				return -1;
+				error_display(0, "Unexpected end of quoted string embedded in sexp!");
+				skip_sexp();
+				return Locked_sexp_false;
 			}
+			// bump past closing quote
+			Mp += (len + 2);
 
-			// check if string variable
-			if ( *(Mp + 1) == SEXP_VARIABLE_CHAR ) {
-				char variable_token[2*TOKEN_LENGTH+2];	// variable_token[contents_token]
-
-				// reduce length by 1 for end \"
-				auto length = len - 1;
-				if (length >= 2*TOKEN_LENGTH+2) {
-					Error(LOCATION, "Variable token %s is too long. Needs to be %d characters or shorter.", Mp, 2*TOKEN_LENGTH+2 - 1);
-					return -1;
-				}
-
-				// start copying after skipping 1st char (i.e. variable char)
-				strncpy(variable_token, Mp + 2, length);
-				variable_token[length] = 0;
-
-				get_sexp_text_for_variable(variable_text, variable_token);
-				node = alloc_sexp(variable_text, (SEXP_ATOM | SEXP_FLAG_VARIABLE), SEXP_ATOM_STRING, -1, -1);
-			} else {
+			// it could be a string variable
+			int sexp_var_index = check_string_for_sexp_variable(startp + 1, len);
+			if (sexp_var_index >= 0) {
+				get_sexp_text_for_variable(token, sexp_var_index);
+				node = alloc_sexp(token, (SEXP_ATOM | SEXP_FLAG_VARIABLE), SEXP_ATOM_STRING, -1, -1);
+			}
+			// it's a regular string
+			else {
 				// token is too long?
 				if (len >= TOKEN_LENGTH) {
-					Error(LOCATION, "Token %s is too long. Needs to be %d characters or shorter.", Mp, TOKEN_LENGTH - 1);
-					return -1;
+					SCP_string long_token(startp + 1, len);
+					error_display(0, "Token is too long. Needs to be %d characters or shorter:\n%s", TOKEN_LENGTH - 1, long_token.c_str());
+					// here we can just truncate; we don't need to return
+					len = TOKEN_LENGTH - 1;
 				}
 
-				strncpy(token, Mp + 1, len);
+				strncpy(token, startp + 1, len);
 				token[len] = 0;
 				node = alloc_sexp(token, SEXP_ATOM, SEXP_ATOM_STRING, -1, -1);
 			}
-
-			// bump past closing \" by 1 char
-			Mp += (len + 2);
-
```

Comment:
Yeah, I went through several iterations before I arrived at this design.  The main issue is that `Mp` needs to point to the following token when `get_sexp()` returns, otherwise you'll have an endless loop.  So I started adding "bumps" before each return, but that started getting ugly because there was a lot of repeated code and yet not all the code bumped the same amount.  And container tokens had to be bumped differently because there was an additional `get_sexp()` in the middle of it.  Rather than introduce a whole lot of fiddly and brittle code, it made more sense to just advance `Mp` as it read the token, while keeping track of where the token started.

Reply:
If it's solely for consistency with the other blocks, then fair enough.